server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -0,0 +1,231 @@
+            .flatMap(blobStore -> blobStore.save(bucketName, data, storagePolicy));
+    }
+
+    @Override
+    public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
+        Preconditions.checkNotNull(data);
+
+        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
+            .flatMap(blobStore -> blobStore.save(bucketName, bufferedInputStream, storagePolicy));
+    }
+
+    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
+        switch (storagePolicy) {
+            case LOW_COST:
+                return Mono.just(lowCostBlobStore);
+            case SIZE_BASED:
+                return largeData.map(isLarge -> {
+                    if (isLarge) {
+                        return lowCostBlobStore;
+                    }
+                    return highPerformanceBlobStore;
+                });
+            case HIGH_PERFORMANCE:
+                return Mono.just(highPerformanceBlobStore);
+            default:
+                throw new RuntimeException("Unknown storage policy: " + storagePolicy);
+        }
+    }
+
+    private boolean isItABigStream(InputStream bufferedData) throws IOException {
+        bufferedData.mark(0);
+        bufferedData.skip(configuration.getSizeThreshold());
+        boolean isItABigStream = bufferedData.read() != -1;
+        bufferedData.reset();
+        return isItABigStream;
+    }
+
+    @Override
+    public BucketName getDefaultBucketName() {
+        Preconditions.checkState(
+            lowCostBlobStore.getDefaultBucketName()
+                .equals(highPerformanceBlobStore.getDefaultBucketName()),
+            "lowCostBlobStore and highPerformanceBlobStore doen't have same defaultBucketName which could lead to " +
+                "unexpected result when interact with other APIs");
+
+        return lowCostBlobStore.getDefaultBucketName();
+    }
+
+    @Override
+    public Mono<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.defer(() -> highPerformanceBlobStore.readBytes(bucketName, blobId))
+            .onErrorResume(this::logAndReturnEmpty)
+            .switchIfEmpty(Mono.defer(() -> lowCostBlobStore.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) {
+        try {
+            return highPerformanceBlobStore.read(bucketName, blobId);
+        } catch (ObjectNotFoundException e) {
+            return lowCostBlobStore.read(bucketName, blobId);
+        } catch (Exception e) {
+            LOGGER.error("Error reading {} {} in {}, falling back to {}", bucketName, blobId, highPerformanceBlobStore, lowCostBlobStore);
+            return lowCostBlobStore.read(bucketName, blobId);
```

Comment:
What is the point of this path ? Shouldn't we have eventually  a retry on the same blobstore ?

Reply:
I must admit I did this to pass a test of "union" blobStore

We need to be careful with retries as they might put undue pressure on already overwhelmed dbs. It may make sense to not loose writes, but failing a read is acceptable IMO.

That being said, in the case of error, trying the other blobStore, just in case, does not hurt...